runtime/altair/src/evm.rs

```diff
@@ -0,0 +1,104 @@
+// Copyright 2023 Centrifuge Foundation (centrifuge.io).
+//
+// This file is part of the Centrifuge chain project.
+// Centrifuge is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version (see http://www.gnu.org/licenses).
+// Centrifuge is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+use cfg_primitives::{MAXIMUM_BLOCK_WEIGHT, NORMAL_DISPATCH_RATIO};
+use frame_support::{
+	parameter_types,
+	traits::FindAuthor,
+	weights::{constants::WEIGHT_REF_TIME_PER_SECOND, Weight},
+	ConsensusEngineId,
+};
+use pallet_evm::{EnsureAddressRoot, EnsureAddressTruncated};
+use runtime_common::evm::{
+	precompile::CentrifugePrecompiles, BaseFeeThreshold, ExpandedAddressMapping,
+};
+use sp_core::{crypto::ByteArray, H160, U256};
+use sp_runtime::Permill;
+use sp_std::marker::PhantomData;
+
+use crate::Aura;
+
+// To create valid Ethereum-compatible blocks, we need a 20-byte
+// "author" for the block. Since that author is purely informational,
+// we do a simple truncation of the 32-byte Substrate author
+pub struct FindAuthorTruncated<F>(PhantomData<F>);
+impl<F: FindAuthor<u32>> FindAuthor<H160> for FindAuthorTruncated<F> {
+	fn find_author<'a, I>(digests: I) -> Option<H160>
+	where
+		I: 'a + IntoIterator<Item = (ConsensusEngineId, &'a [u8])>,
+	{
+		if let Some(author_index) = F::find_author(digests) {
+			let authority_id = Aura::authorities()[author_index as usize].clone();
+			return Some(H160::from_slice(&authority_id.to_raw_vec()[4..24]));
+		}
+		None
+	}
+}
```

Comment:
This kind of runtime-agnostic trait impl could to to `runtime-commons` and just be referenced in the different runtimes instead of duplicated in all of them 🧹

Reply:
I think it needs to be made generic over the Runtime (that `Aura` in there is really `pallet_aura::Pallet::<Runtime>`), but indeed could be made properly runtime agnostic. Worth doing for sure

Reply:
Ah, my bad, I didn't spot the `Aura` bit 🙃 I don't know how to make stuff generic over a runtime when referencing a specific pallet from that runtime tho.

Reply:
We could add a `Get<Vec<Authorities>>` and impl that for Aura in the actual runtime. But feel like this makes it more opaque than actually easier to understand.